pkg/clustermesh/common/namespacewatcher/watcher.go

```diff
@@ -0,0 +1,270 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package namespacewatcher
+
+import (
+	"log/slog"
+	"strings"
+
+	"github.com/spf13/pflag"
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	"github.com/cilium/cilium/pkg/annotation"
+	"github.com/cilium/cilium/pkg/k8s/resource"
+	slim_corev1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/api/core/v1"
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+)
+
+// GlobalNamespaceTracker tracks which namespaces are marked as global for ClusterMesh export.
+type GlobalNamespaceTracker interface {
+	// IsGlobalNamespace returns true if the given namespace should be exported globally.
+	// Returns true by default for backwards compatibility if no specific configuration exists.
+	IsGlobalNamespace(namespace string) bool
+
+	// GetGlobalNamespaces returns the set of namespaces that are currently marked as global.
+	GetGlobalNamespaces() sets.Set[string]
+
+	// RegisterProcessor registers a processor that will be notified when namespace status changes
+	RegisterProcessor(processor NamespaceProcessor)
+
+	// IsGlobalService checks if a service is global considering both
+	// the service annotation and namespace filtering when active.
+	// Expectation is this function will be invoked with
+	// "github.com/cilium/cilium/pkg/loadbalancer".Service
+	IsGlobalService(obj interface {
+		GetAnnotations() map[string]string
+		GetNamespace() string
+	}) bool
+
+	// IsSharedGlobalService checks if a service is both global and shared
+	IsSharedGlobalService(obj interface {
+		GetAnnotations() map[string]string
+		GetNamespace() string
+	}) bool
```

Comment:
I'd argue that it is not responsibility of this module to know about services, and how they are defined as global/shared. I'd keep things simple, and just add the namespace check at the callers as appropriate.